prototype walker

doug-q opened the pull request.

doug-q commented:
Needs docs. Feel free to suggest better names.

acl-cqc commented on src/walker.rs:
```diff
@@ -0,0 +1,246 @@
+                    for cb in self.mut_callbacks(order).iter_mut() {
+                        // this clone is unfortunate, to avoid this we would
+                        // need a TryInto variant like:
+                        // try_into(&O) -> Option<&T>
```
I bet there is some way with `enum_dispatch` but I'll leave that to someone Rustier than me.... @aborgna-q ? ;)

A clone is no big deal, I'm happy with that

acl-cqc reviewed (commented):
I love this stack of pre/post visit WorkItems! :-). I think my biggest concern is with CFG nodes - I gather that is the intention, so is the idea to walk only the single basic block inside the CFG inside the function definition that get from Guppy, or what?

doug-q commented on src/walker.rs:
```diff
@@ -0,0 +1,246 @@
+
+    pub fn walk(&mut self, mut t: T) -> Result<T, E> {
+        // We intentionally avoid recursion so that we can robustly accept very deep hugrs
+        let mut worklist = vec![WorkItem::Visit(self.hugr.root())];
```
Do you mean the definition? that would be better.

doug-q commented on src/walker.rs:
```diff
@@ -0,0 +1,246 @@
+                    // the input node. (For example, LoadConstant nodes may not
+                    // be reachable from the input node). So we do a second
+                    // unordered traversal afterwards.
+                    if let Some([input, _]) = self.hugr.get_io(n) {
```
CFG and Module nodes work fine, they are caught be the second traversal. There is a test on Modules

acl-cqc replied:
Ah I see got it. I hadn't lined up the indenting correctly / realized the closing `}`. In that case this all looks good, but would be good to mention in the comments

acl-cqc replied:
Maybe just take "So we do a second unordered traversal aftewards", put on a new line, and add "this covers non-dataflow sibling graphs too"?

doug-q commented on src/walker.rs:
```diff
@@ -0,0 +1,246 @@
+                    // We intend to only visit direct children.
+                    //
+                    // If the nodes children form a dataflow sibling graph we
+                    // visit them in post dfs order. This traversal is not
```
I agree a better comment is needed. Post DFS is reverse topological sort, but we push them to a stack so they are visited in reverse Post DFS order.

doug-q commented on src/walker.rs:
```diff
@@ -0,0 +1,246 @@
+        self
+    }
+
+    pub fn walk(&mut self, mut t: T) -> Result<T, E> {
```
Because the callbacks need access to the hugr. In the current design the hugr is fixed for a given Walker, so the callbacks can capture the hugr. Passing the hugr here would mean that the Walker could be invoked on many hugrs, so the callbacks couldn't capture it. The hugr would need to be threaded through the callbacks. This can be done but I think it's less ergonomic.

ss2165 reviewed (commented):
Looks good but my main question is why the behaviour is "walk the hugr, and visit nodes of a certain type, doing something to some accumulated data when you visit" (if I've understood correctly).

It seems more useful in general to do "walk the hugr, returning an iterator" and then you can call a normal filter or whatever on the iterator, and for callbacks a more general input would be just the node (maybe the + an immutable hugrview?) and I can make arbitrary querires on the hugr to decide what to do to the accumulated data, like queries about connectivity that I can't get from just the node weight.

I might be missing something about the functionality here - if so it would be nice to have tests which demonstrate the kinds of use case I'm talking about. (Suggested test - walk the hugr and record a hashmap keyed by node, only for LeafOps, with the value being the degree of the node, including counting each edge of a multiport).

doug-q commented:
> Looks good but my main question is why the behaviour is "walk the hugr, and visit nodes of a certain type, doing something to some accumulated data when you visit" (if I've understood correctly).

The behaviour is this way so that one has the option of maintaining a context as you descend/ascend the heirarchy. See the new pretty_printer test where the indent level is increased once for each level of nesting.

> It seems more useful in general to do "walk the hugr, returning an iterator" and then you can call a normal filter or whatever on the iterator, and for callbacks a more general input would be just the node (maybe the + an immutable hugrview?) and I can make arbitrary querires on the hugr to decide what to do to the accumulated data, like queries about connectivity that I can't get from just the node weight.

I had thought that this was general enough to implement that iterator, but on reflection (and see hugr_walk_iter()) this is not good enough, because it's not lazy. walk should be changed to return an Iterator and the implementation of walk moved into that Iterator implementation.

I've changed the general form of callback to not take the OpType. I've not changed the signature of visit, the
> 
> 
> I might be missing something about the functionality here - if so it would be nice to have tests which demonstrate the kinds of use case I'm talking about. (Suggested test - walk the hugr and record a hashmap keyed by node, only for LeafOps, with the value being the degree of the node, including counting each edge of a multiport).
  
 I've added this test(mostly, I need to add an actual hugr for it to run on), as well as a pretty printing test that shows the usefulness of maintaining context as you traverse the heirarchy. I've also added a find function, demonstrating early return.
